Sources/Core/Shared/GroupOperation.swift

```diff
@@ -339,3 +448,117 @@ public struct WillAddChildObserver: GroupOperationWillAddChildObserver {
         didAttachToOperation?(operation: operation)
     }
 }
+
+private extension GroupOperation {
+    /**
+     The group operation handles thread-safe addition of operations by utilizing two final operations:
+     - a CanFinishOperation which manages handling GroupOperation internal state and has every child
+       operation as a dependency
+     - a finishingOperation, which has the CanFinishOperation as a dependency
+
+     The purpose of this is to handle the possibility that GroupOperation.addOperation() or
+     GroupOperation.queue.addOperation() are called right after all current child operations have
+     completed (i.e. after the CanFinishOperation has been set to ready), but *prior* to being able
+     to process that the GroupOperation is finishing (i.e. prior to the CanFinishOperation executing and
+     acquiring the GroupOperation.groupFinishLock to set state).
+     */
+    private class CanFinishOperation: NSOperation {
+        private weak var parent: GroupOperation?
+        private var _finished = false
+
+        init(parentGroupOperation: GroupOperation) {
+            self.parent = parentGroupOperation
+            super.init()
+        }
+        override func start() {
+            main()
+        }
+        override func main() {
+            execute()
+        }
+        func execute() {
+            if let parent = parent {
+
+                // All operations that were added as a side-effect of anything up to
+                // WillFinishObservers of prior operations should have been executed.
+                //
+                // Handle an edge case caused by concurrent calls to GroupOperation.addOperations()
+
+                let isWaiting = parent.groupFinishLock.withCriticalScope { () -> Bool in
+
+                    // Is anything currently adding operations?
+                    guard dispatch_group_wait(parent.isAddingOperationsGroup, DISPATCH_TIME_NOW) == 0 else {
+                        // Operations are actively being added to the group
+                        // Wait for this to complete before proceeding.
+                        dispatch_group_notify(parent.isAddingOperationsGroup,
+                            dispatch_get_global_queue(NSQualityOfServiceToDispatchQOS(self.qualityOfService), 0),
+                            { [unowned self] in
+                            self.execute()
+                        })
+                        return true
+                    }
+
+                    // Check whether new operations were added prior to the lock
+                    // by checking for child operations that are not finished.
+
+                    let nonFinishedOperations = parent.operations.filter({ !$0.finished })
+                    if !nonFinishedOperations.isEmpty {
```

Comment:
I think here, from a readability point of view - it's best to try to express things to avoid double negatives. It just requires a little bit less mental jumping when reading through code. Suggest we rename this constant to `activeOperations`. This was, `if !activeOperation.isEmpty { } or `if activeOperations.count > 0 { }` is clearer what the logic is expressing.